crates/holochain/src/core/ribosome.rs

```diff
@@ -9,7 +9,48 @@
 // documentation, and there seems to be no way to add docs to it after the fact
 #[allow(missing_docs)]
 pub mod error;
+
+/// How to version guest callbacks.
+/// See `genesis_self_check` for an example.
+///
+/// - Create unversioned structs in the root of the callback module
+///   - Invocation, result, host access
+///   - The unversioned structs should thinly wrap all their versioned structs
+/// - Create versioned submodules for the callback
+///   - In these, create versioned structs for the unversioned structs
+///   - Write/keep all tests for the versioned copies of the callbacks, test
+///     wasms can expose externs directly without the macros for explicit
+///     legacy identities if needed.
+/// - On the ribosome make sure the trait uses the unversioned struct
+///   - Inside the callback method loop over the versioned callbacks, and
+///     dispatch each that is found in the wasm
+///   - Figure out how to merge/handle results if multiple versions of a callback
+///     are found in the target wasm
+/// - The ribosome method caller will now be forced by types to provide the
+///   unversioned struct, which means they cannot forget to provide and dispatch
+///   everything required for each version
+/// - Update the `map_extern` macro so that the unversioned name of the callback
+///   maps to the latest version of the callback, e.g. `genesis_self_check` is
+///   rewritten to `genesis_self_check_2` at the time of writing
+///   - This has the effect of newly compiled wasms implementing the callback
+///     that is newest when they compile, without polluting the unversioned
+///     callback, which is effectively legacy/deprecated behaviour to call it
+///     directly.
```

Comment:
It's certainly a lot to remember! Judging from this PR it looks like any time we version a new fn it will require >500 lines changed. I wonder if there's any low hanging fruit in the way of automating some of this or at least putting it in the type system so that the compiler will force us to do the update properly?

Reply:
@maackle the compiler does already force a lot of this, although i can't say it's 100%, the documentation is just another safety net

Reply:
@maackle but yes, it's a pain to modify interfaces, that's why i was pushing that long convo we had about minimising the surface area of callback signatures and shift the focus to HDK functions for happ devs to get at data. HDK functions are significantly easier to version than callbacks, partly because they are always singular when called, but callbacks need a cross-version "fan out and aggregate" strategy by the host caller.